### DISCUSSION TITLE
SBC settings

### DISCUSSION THREAD

**Discussion by litezxc** [pschatzmann/ESP32-A2DP#639] [Q&A] [answered]
Hello,

I noticed that the BluetoothSink library only includes standard settings for the SBC codec. I was wondering if it would be possible to integrate settings for the SBC codec from the additional library, Arduino-libSBC.

Currently, I am working on a project where sound is transmitted from a PC to an ESP32 module and output to a DAC. I have a preliminary code in place for this setup:

#include "BluetoothA2DPSink.h"

#include "SBC.h"  // Include the libsbc library

BluetoothA2DPSink a2dp_sink;

// Define I2S pins
#define I2S_BCK  26 // BCK pin
#define I2S_WS   25 // WS pin
#define I2S_DO   22 // DATA pin

void setup() {
    Serial.begin(115200);

    // Initialize I2S
    i2s_config_t i2s_config = {
        .mode = (i2s_mode_t)(I2S_MODE_MASTER | I2S_MODE_TX), // Master mode and transmission
        .sample_rate = 48000,                                 // Sample rate in Hz
        .bits_per_sample = I2S_BITS_PER_SAMPLE_16BIT,       // Bits per sample
        .channel_format = I2S_CHANNEL_FMT_RIGHT_LEFT,       // Stereo channels
        .communication_format = I2S_COMM_FORMAT_I2S,        // I2S format
        .dma_buf_count = 8,                                  // Number of DMA buffers
        .dma_buf_len = 1024                                   // Length of DMA buffer
    };

    // Install I2S driver
    i2s_driver_install(I2S_NUM_0, &i2s_config, 0, NULL);

    // Set I2S pins
    i2s_pin_config_t pin_config = {
        .bck_io_num = I2S_BCK,
        .ws_io_num = I2S_WS,
        .data_out_num = I2S_DO,
        .data_in_num = I2S_PIN_NO_CHANGE // Not using input data
    };
    i2s_set_pin(I2S_NUM_0, &pin_config);

    // Initialize Bluetooth A2DP Sink
    a2dp_sink.start("ESP32_A2DP_Sink"); // Device name

    // Set up SBC codec using libsbc
    SBCEncoder sbcEncoder; // Create SBCEncoder object
    sbcEncoder.setNumOfSubbands(8); // Set number of subbands
    sbcEncoder.setBlockLength(16);   // Set number of blocks
    sbcEncoder.setBitPool(80);        // Set bit pool
    sbcEncoder.setAllocationMethod(SBC_AM_LOUDNESS); // Set allocation method
    
    // Use the SBC encoder in the A2DP Sink
    a2dp_sink.setSBCEncoder(&sbcEncoder); // Assuming setSBCEncoder accepts a pointer to SBCEncoder

    Serial.println("Bluetooth A2DP Sink has started.");
}

void loop() {
    // Your main code can go here
}

Could you please provide guidance on what needs to be done to make this integration work? Specifically, how can I use configurations from Arduino-libSBC to customize the SBC settings in the BluetoothSink library?

Thank you for your assistance!

**Comment by pschatzmann** [comment, marked as answer]
No, the [ESP IDF API](https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/bluetooth/esp_a2dp.html) does not provide any such functionality

**Reply by litezxc** [reply]
Is it possible to do this in any other environment? Could you provide recommendations?

**Reply by pschatzmann** [reply]
If you search in Github, you will find some custom adapted (I guess old version of) IDF which supports different codecs.

But I guess it would be better to submit a request to Espressif  to provide this functionality in the future....

**Reply by litezxc** [reply]
Thank you for your responses, I really appreciate it
